openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,435 @@
+	}
+	if a.nodeListSynced == nil {
+		return fmt.Errorf("%s plugin needs a node lister synced", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	workloadAnnotations := getWorkloadAnnotations(pod.Annotations)
+
+	// strip any resource annotations specified by a user
+	stripResourcesAnnotations(pod.Annotations)
+	// strip any workload annotation to prevent from underlying components(CRI-O, kubelet) to apply any changes
+	// according to the workload annotation
+	stripWorkloadAnnotations(pod.Annotations)
+
+	// more than one workload annotation exists under the pod and we do not support different workload types
+	// under the same pod
+	if len(workloadAnnotations) > 1 {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningMultipleWorkloadAnnotations, workloadAnnotations)
+		return nil
+	}
+
+	workloadType, err := getWorkloadType(workloadAnnotations)
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadType, err)
+		return nil
+	}
+
+	// no workload annotation is available
+	if workloadType == "" {
+		return nil
+	}
+
+	workloadAnnotation := fmt.Sprintf("%s%s", podWorkloadAnnotationPrefix, workloadType)
+	// TODO: currently we do not check the effect, but it can be useful in the future
+	_, err = getWorkloadAnnotationEffect(workloadAnnotations[workloadAnnotation])
+	if err != nil {
+		pod.Annotations[workloadAdmissionWarning] = fmt.Sprintf(admissionWarningFailedGetWorkloadAnnotationEffect, err)
+		return nil
+	}
+
+	if !a.waitForSyncedStore(time.After(timeToWaitForCacheSync)) {
+		return admission.NewForbidden(attr, fmt.Errorf("%s node and namespace caches not synchronized", PluginName))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return err
+		}
+
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if !isNamespaceAllowed(ns.Annotations, workloadType) {
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		return admission.NewForbidden(attr, err) // can happen due to informer latency
+	}
+
+	if len(nodes) == 0 {
+		return nil
+	}
+
+	// TODO: currently we supports only SNO use case because we have not yet worked out the best approach to determining whether the feature
+	// should be on or off in a multi-node cluster, and computing that state incorrectly could lead to breaking running clusters.
+	if len(nodes) > 1 {
+		return nil
```

Comment:
This seems like a weak guarantee during installation.  Allowing pods that don't have intent fulfilled is naughty.

Reply:
agree, but probably this one + check for manager resource should be good enough for, but we definitely will improve it under 4.9.

Reply:
Please set the `workloadAdmissionWarning` indicating the intent was ignored.

Reply:
@deads2k It going in conflict with @sttts comment https://github.com/openshift/kubernetes/pull/632#discussion_r611673297, can you please decide what behavior is acceptable.

Reply:
Not sure about the exact context of my former comment. Basically don't add warnings to all pods, i.e. especially if they don't fall under the conditions checked here (= if they don't have the workload annotations). But adding a warning for a pod (with annotations) when the specified intent is ignored it good (as @deads2k suggests).

Reply:
done